src/mlpack/methods/bayesian_ridge/bayesian_ridge.cpp

```diff
@@ -0,0 +1,279 @@
+{
+  Timer::Start("bayesian_ridge_regression");
+
+  arma::mat phi;
+  arma::rowvec t;
+  arma::colvec eigval;
+  arma::mat eigvec;
+  arma::colvec eigvali;
+
+  // Preprocess the data. Center and scale.
+  CenterScaleData(data,
+                  responses,
+                  centerData,
+                  scaleData,
+                  phi,
+                  t,
+                  dataOffset,
+                  dataScale,
+                  responsesOffset);
+
+  // Compute this quantities once and for all.
+  const arma::colvec vecphitT = phi * t.t();
+  const arma::mat phiphiT =  phi * phi.t();
+
+  arma::eig_sym(eigval, eigvec, phiphiT);
+
+  // Detect singular matrix.
+  if (eigval[0] < 1e-8)
+  {
+    Log::Warn << "Singular matrix. Two lines or more are colinear."
+              <<  std::endl;
+    return -1;
+  }
+
+  unsigned short p = data.n_rows, n = data.n_cols;
+  // Initialize the hyperparameters and
+  // begin with an infinitely broad prior.
+  alpha = 1e-6;
+  beta =  1 / (var(t) * 0.1);
+
+  unsigned short nIterMax = 50;
+  unsigned short i = 0;
+  double deltaAlpha = 1, deltaBeta = 1, crit = 1;
+  arma::mat matA = arma::eye<arma::mat>(p, p);
+
+  while ((crit > tol) && (i < nIterMax))
+  {
+    deltaAlpha = -alpha;
+    deltaBeta = -beta;
+
+    // Compute the posterior statistics.
+    // with inv()
+    matA.diag().fill(alpha);
+    // inv is used instead of solve because we need the covariance matrix to
+    // compute the prediction uncertainties. If solve is used, matCovariance
+    // must be comptuted at the end of the loop.
+    matCovariance = inv_sympd(matA + phiphiT * beta);
+    omega = (matCovariance * vecphitT) * beta;
+
+    // // with solve()
+    // matA.diag().fill(alpha/ beta);
+    // omega = solve(matA + phiphiT, vecphitT);
+
+    // Update alpha.
+    eigvali = eigval * beta;
+    gamma = sum(eigvali / (alpha + eigvali));
+    alpha = gamma / dot(omega.t(), omega);
+
+    // Update beta.
+    const arma::rowvec temp = t - omega.t() * phi;
+    beta = (n - gamma) / dot(temp, temp);
+
+    // Comptute the stopping criterion.
+    deltaAlpha += alpha;
+    deltaBeta += beta;
+    crit = abs(deltaAlpha/alpha + deltaBeta/beta);
+    i++;
+  }
+  Timer::Stop("bayesian_ridge_regression");
+  return Rmse(data, responses);
+}
+
+void BayesianRidge::Predict(const arma::mat& points,
+                            arma::rowvec& predictions) const
+{
+  arma::mat X = points;
+
+  // Center and scaleData the points before applying the model
+  X.each_col() -= dataOffset;
+  X.each_col() /= dataScale;
+  predictions = omega.t() * X + responsesOffset;
+}
+
+void BayesianRidge::Predict(const arma::mat& points,
+                            arma::rowvec& predictions,
+                            arma::rowvec& std) const
+{
+  arma::mat X = points;
+
+  // Center and scaleData the points before applying the model.
+  X.each_col() -= dataOffset;
+  X.each_col() /= dataScale;
+  predictions = omega.t() * X + responsesOffset;
```

Comment:
I have inlined the calculation of the uncertainties. X.col(i) are not used anymore. Because I needed to use the centered and scaled design matrix X several times, I kept the copy. Let me know, if there is something better.